In [85]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
sns.set()

from sklearn.linear_model import LinearRegression

# **Simple linear regression**

**Load data**

In [86]:
data = pd.read_csv('sample_data/1.01.+Simple+linear+regression.csv')
data.head()

SAT   GPA
0  1714  2.40
1  1664  2.52
2  1760  2.54
3  1685  2.74
4  1693  2.83

# **Create the regression**

**Declear the dependent and independent variables**

In [87]:
x = data['SAT']
y = data['GPA']
# 我们的x称为输入或特征, 而Y称为输出或目标｡

在监督学习中, 我们有输入和目标｡ 输入是我们用来预测结果的特征｡ 在我们的例子中, SAT分数｡

通常, 目标是对应于给定SAT分数的历史值｡ 实际上, 我们的机器学习算法将找到线性回归模型的最优系数｡ 那是在给出SAT分数的时候｡ 最好预测学生的GPA｡

        y = output(traget)         <--         x = input(feature)
y is dependent variable

x is independent variable
       

In [88]:
x.shape

(84,)

shape() is return the index, number of the element

In [89]:
y.shape

(84,)

这里的 x 是一维数组（shape = (n,)）。
但是 scikit-learn 的回归模LinearRegression() 要求输入是二维的 (shape = (n_samples, n_features))：

n_samples = 数据点数量（行数）
n_features = 特征数量（列数）

In [90]:
x_matrix = x.values.reshape(-1,1)
x_matrix.shape

# -1 表示“自动计算行数”，会变成样本数。
# 1 表示只有一个特征。

(84, 1)

reshape(-1,1) 是把一维向量变成二维矩阵（因为 sklearn 要求输入必须是二维 [n_samples, n_features] 的形式）。

例如： 如果 x_scaled[:,0] = [1.2, 0.3, -0.7] 则 reshape 后变成：

[[ 1.2], [ 0.3], [-0.7]]

**Regression itself**

In [91]:
reg = LinearRegression()

In [92]:
reg.fit(x_matrix,y)
# fit(input,target)

LinearRegression()

**R-squared**

In [93]:
reg.score(x_matrix,y)
# R-squared

0.40600391479679754

**Coefficients**

In [94]:
reg.coef_
# SAT coef

array([0.00165569])

**Intercept**

In [95]:
reg.intercept_
# const coef

np.float64(0.2750402996602799)

**Making predictions**

In [96]:
reg.predict(np.array([[1740]]))

array([3.15593751])

In [97]:
new_data = pd.DataFrame(data=[1740,1760],columns=['SAT'])
new_data

SAT
0  1740
1  1760

In [98]:
reg.predict(new_data)

/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2732: UserWarning: X has feature names, but LinearRegression was fitted without feature names
  warnings.warn(


array([3.15593751, 3.18905127])

In [99]:
new_data['Predicted_GPA'] = reg.predict(new_data)
new_data

/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2732: UserWarning: X has feature names, but LinearRegression was fitted without feature names
  warnings.warn(


SAT  Predicted_GPA
0  1740       3.155938
1  1760       3.189051

通过regression()能算出所有的数据ex：R-squared, SAT coef, const coef

然后用predict()就会用公式ex: GPA=−1.41+0.0023⋅SAT 算出对应的 GPA 预测值。

# **Multiple linear regression**

In [100]:
dataMul = pd.read_csv('sample_data/1.02.+Multiple+linear+regression.csv')
dataMul.head()

SAT  Rand 1,2,3   GPA
0  1714           1  2.40
1  1664           3  2.52
2  1760           3  2.54
3  1685           3  2.74
4  1693           2  2.83

In [101]:
dataMul.describe()

SAT  Rand 1,2,3        GPA
count    84.000000   84.000000  84.000000
mean   1845.273810    2.059524   3.330238
std     104.530661    0.855192   0.271617
min    1634.000000    1.000000   2.400000
25%    1772.000000    1.000000   3.190000
50%    1846.000000    2.000000   3.380000
75%    1934.000000    3.000000   3.502500
max    2050.000000    3.000000   3.810000

# **Create multiple linear regression**

**Declear the dependent and independent variable**

In [102]:
xMul = dataMul[['SAT','Rand 1,2,3']]
yMul = dataMul['GPA']

**Regression itself**

In [103]:
reg = LinearRegression()
reg.fit(xMul,yMul)

LinearRegression()

**SAT coef, Rand1,2,3 coef**

In [104]:
reg.coef_

array([ 0.00165354, -0.00826982])

**const coef**

In [105]:
reg.intercept_

np.float64(0.29603261264909486)

# **Calculating the R-squared**

**R-squared：决定系数，表示模型能解释多少 GPA 的波动**

In [106]:
reg.score(xMul,yMul)

0.4066811952814283

SAT 能解释 GPA 波动的 40.6%，剩下的 59.4% 差异，需要其他因素来解释（比如努力程度、课程选择、学习环境等）。

**Formula for Adjusted R^2**

R²adj = 1 - [(1 - R²)(n - 1) / (n - p - 1)]

In [107]:
xMul.shape

(84, 2)

In [108]:
RSquared = reg.score(xMul,yMul)

n = xMul.shape[0]
p = xMul.shape[1]

adjusted_r2 = 1-(1 - RSquared)(n - 1)/(n - p - 1)
adjusted_r2

0.3920313482513401

# **Feature selection**

我们已经创建了一个multiple regression, 并计算了其调整后的R-squared｡

R-squared的值表明两个变量中的一个并不能增加回归的解释力｡ (SAT和rand1,2,3分别获得的R-squared, rand1,2,3的R-squared表明rand1,2,3这个数据模型没有用)

因此, 如果我们能识别出是哪一个并将其删除, 我们就可以进一步简化模型｡

In [109]:
from sklearn.feature_selection import f_regression

In [110]:
f_regression(xMul,yMul)

(array([56.04804786,  0.17558437]), array([7.19951844e-11, 6.76291372e-01]))

In [111]:
p_values = f_regression(xMul,yMul)[1]
p_values

array([7.19951844e-11, 6.76291372e-01])

In [112]:
p_values.round(3)

array([0.   , 0.676])

**F-statistic 和 p-value 的关系**

F-statistic 测试的是：这个特征对 y 的解释力是否显著大于 0。

p-value 是根据 F-statistic 算出来的显著性概率。

**如果特征:**

p-value 很小（比如 < 0.05），说明它对 y 的影响显著 → 这个特征重要

p-value 大（比如 > 0.5），说明它对 y 的解释力不显著 → 可以考虑剔除

# **Creating a summary table**

In [113]:
reg_summary = pd.DataFrame(data = xMul.columns.values, columns=['Features'])
reg_summary

Features
0         SAT
1  Rand 1,2,3

In [114]:
reg_summary['coefficients'] = reg.coef_
reg_summary['p-values'] = p_values.round(3)

In [115]:
reg_summary

Features  coefficients  p-values
0         SAT      0.001654     0.000
1  Rand 1,2,3     -0.008270     0.676

**coefficients：回归系数，代表这个特征对目标变量（如 GPA）的影响大小和方向。**

SAT 的系数 0.001654 → SAT 每提高 1 分，GPA 平均提高 0.001654。

Rand 的系数 -0.008270 → 随机数对 GPA 的影响是负的，但很小。

**p-values：检验该特征对模型是否有显著贡献（假设检验）。**

SAT 的 p = 0.000 → 显著，说明 SAT 对 GPA 的预测有意义。

Rand 的 p = 0.676 → 不显著，说明随机数和 GPA 没关系。

**coefficients和p-values都说明Rand1,2,3没有用**

# **Standardization**

fit()：学到每一列的均值和标准差

transform()：应用标准化，把原始数据变成：

x′=（x−μ​）/ σ

这样每个特征的均值=0，标准差=1。

In [116]:
from sklearn.preprocessing import StandardScaler

In [117]:
scaler = StandardScaler()
# 1. 定义一个标准化工具

In [118]:
scaler.fit(xMul)
# 2. 计算每个特征的均值和标准差

StandardScaler()

In [119]:
x_scaled = scaler.transform(xMul)
# 3. 用均值和标准差把数据转换

In [120]:
x_scaled

array([[-1.26338288, -1.24637147],
       [-1.74458431,  1.10632974],
       [-0.82067757,  1.10632974],
       [-1.54247971,  1.10632974],
       [-1.46548748, -0.07002087],
       [-1.68684014, -1.24637147],
       [-0.78218146, -0.07002087],
       [-0.78218146, -1.24637147],
       [-0.51270866, -0.07002087],
       [ 0.04548499,  1.10632974],
       [-1.06127829,  1.10632974],
       [-0.67631715, -0.07002087],
       [-1.06127829, -1.24637147],
       [-1.28263094,  1.10632974],
       [-0.6955652 , -0.07002087],
       [ 0.25721362, -0.07002087],
       [-0.86879772,  1.10632974],
       [-1.64834403, -0.07002087],
       [-0.03150724,  1.10632974],
       [-0.57045283,  1.10632974],
       [-0.81105355,  1.10632974],
       [-1.18639066,  1.10632974],
       [-1.75420834,  1.10632974],
       [-1.52323165, -1.24637147],
       [ 1.23886453, -1.24637147],
       [-0.18549169, -1.24637147],
       [-0.5608288 , -1.24637147],
       [-0.23361183,  1.10632974],
       [ 1.68156984,

# **Regression with scaled features**

In [121]:
reg = LinearRegression()
reg.fit(x_scaled,y)

LinearRegression()

In [122]:
reg.coef_

array([ 0.17181389, -0.00703007])

In [123]:
reg.intercept_

np.float64(3.330238095238095)

# **Creating a summary table**

In [124]:
reg_summary = pd.DataFrame([['Bias'],['SAT'],['Rand 1,2,3']], columns=['Features'])
reg_summary['weights'] = reg.intercept_, reg.coef_[0], reg.coef_[1]

In [125]:
reg_summary

Features   weights
0        Bias  3.330238
1         SAT  0.171814
2  Rand 1,2,3 -0.007030

SAT 的 weight = 0.0016 → SAT 每提高 1 分，GPA 提高 0.0016

Rand 的 weight = -0.008 → Rand 越大，GPA 越低（Rand对我们的输出几乎没有贡献）

bias = 0.27 → 截距，当 SAT 和 Rand=0 时，预测 GPA=0.27

# **Making predictions with the standardized coefficients (weights)**

这里是一个小的example，创建一个dataframe，这里只有两个data，fit()：学到每一列的均值和标准差，transform()：应用标准化，把原始数据变成：

x′=（x−μ​）/ σ

这样每个特征的均值=0，标准差=1。

在standardized后用predict()来算出预算值。
y = b+βSAT​ * SAT + βRand * Rand

In [135]:
new_data = pd.DataFrame(data=[[1700,2],[1800,1]],columns=['SAT','Rand 1,2,3'])
new_data

SAT  Rand 1,2,3
0  1700           2
1  1800           1

In [136]:
reg.predict(new_data)

/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2732: UserWarning: X has feature names, but LinearRegression was fitted without feature names
  warnings.warn(


array([295.39979563, 312.58821497])

In [137]:
new_data_scaled = scaler.transform(new_data)
new_data_scaled

array([[-1.39811928, -0.07002087],
       [-0.43571643, -1.24637147]])

In [138]:
reg.predict(new_data_scaled)
# 算出预测的 GPA。

array([3.09051403, 3.26413803])

# **What if we removed the 'Random 1,2,3' variable?**

In [148]:
reg_simple = LinearRegression()
x_simple_matrix = x_scaled[:,0].reshape(-1,1)
reg_simple.fit(x_simple_matrix,y)

LinearRegression()

x_scaled = [SAT,Rand 1,2,3]

x_scaled[:,0]从 x_scaled 中取出第 0 列特征（比如 SAT 分数）

reshape(-1,1) 是把一维向量变成二维矩阵（因为 sklearn 要求输入必须是二维 [n_samples, n_features] 的形式）。

例如：
如果 x_scaled[:,0] = [1.2, 0.3, -0.7]
则 reshape 后变成：

[[ 1.2],
 [ 0.3],
 [-0.7]]

reg_simple.fit(x_simple_matrix, y)

用 x_simple_matrix (特征) 和 y (目标变量，比如 GPA) 去训练回归模型。

这一步会得到 斜率 (coef_) 和截距 (intercept_)。

In [150]:
d = new_data_scaled[:,0].reshape(-1,1)
d

array([[-1.39811928],
       [-0.43571643]])

In [145]:
reg_simple.predict(new_data_scaled[:,0].reshape(-1,1))

array([3.08970998, 3.25527879])

new_data_scaled[:,0].reshape(-1,1) = [[-1.39811928],[-0.43571643]] (这个是小的example中的SAT)

predict() 的输入也必须是 二维矩阵

用训练好的模型对新的数据（new_data_scaled 的第 0 列）做预测。

返回的就是通过公式y = b+βSAT​ * SAT预测的目标值（比如 GPA）。